### PR TITLE
Adding version detection function to new build

### PR DESCRIPTION
**Pull request by sdavtaker**
- Setting up project version and setting plan for next steps

*Issue #, if available:*

*Description of changes:*
Since this is the first new function we introducing in the build system and a few things are different to those in the LEGACY implementation I will comment on most of its lines in this PR to highlight the changes and their reasons.
This particular PR is the best place to discuss about those decisions, that may seem small here, but will affect the whole build system code style and practices long term. 

*Check all that applies:*
- [x] Did a review by yourself.
- [x] Checked if this PR will _not_ introduce cross-platform inconsistent behaviour.
- [x] Checked if this PR would require a ReadMe/Wiki update.
         README changes coming in a follow up PR after the build system is completely refactored. 

Check which platforms you have built SDK on to verify the correctness of this PR.
- [x] Linux
- [ ] Windows
- [ ] Android
- [x] MacOS
- [ ] IOS
- [ ] Other Platforms

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,17 +20,46 @@ else ()
     message(STATUS "Building with 1.10 CMake scripts.")
     string(CONCAT DESCRIPTION_STRING "The AWS SDK for C++ provides a modern C++ (standard version C++11 or later) "
             "interface for Amazon Web Services (AWS).")
+
+    if (NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
+        message(STATUS "Setting build type to 'RelWithDebInfo' as none was specified.")
+        set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE
+                STRING "Choose the type of build." FORCE)
+        # cmake-gui helper
+        set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
```
Comment (sdavtaker) [inline CMakeLists.txt:29]: This helps gui tools as cmake to show only a list of valid options to configure the CMAKE_BUILD_TYPE variable in place of having a free form string as input.

File: CMakeLists.txt [hunk 2]
```diff
@@ -20,17 +20,46 @@ else ()
...
+    endif ()
+
+    list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")
+    include(sdk_versioning)
```
Comment (sdavtaker) [inline CMakeLists.txt:33]: The module has no side effects, it contains only function/macro definitions.

File: CMakeLists.txt [hunk 3]
```diff
@@ -20,17 +20,46 @@ else ()
...
+    message(STATUS "Building core library.")
+    message(STATUS "Building core library tests.")
+    message(STATUS "Add support for static analysis.")
+    message(STATUS "Building client libraries.")
+    message(STATUS "Setting up tests.")
+    message(STATUS "Setting up installation.")
+    message(STATUS "Setting up packaging. ")
+    message(STATUS "Setting up docs.")
+    message(STATUS "Add support support for SDK flags.")
+    message(STATUS "Add previously available options.")
```
Comment (sdavtaker) [inline CMakeLists.txt:63]: TODO items

File: cmake/sdk_versioning.cmake [hunk 4]
```diff
@@ -0,0 +1,20 @@
+include_guard()
+
+function(obtain_project_version resultVar)
```
Comment (sdavtaker) [inline cmake/sdk_versioning.cmake:3]: resultVar allows caller to define whats the variable that will be overridden at the parent scope

File: cmake/sdk_versioning.cmake [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    find_package(Git QUIET)
+    if (GIT_FOUND)
+        execute_process(COMMAND ${GIT_EXECUTABLE} --git-dir=${CMAKE_CURRENT_SOURCE_DIR}/.git describe --abbrev=0 --tags
+                WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+                OUTPUT_VARIABLE VERSION_STRING
+                OUTPUT_STRIP_TRAILING_WHITESPACE)
+    endif ()
```
Comment (sdavtaker) [inline cmake/sdk_versioning.cmake:10]: Trying to infer version from git tags if available.

File: cmake/sdk_versioning.cmake [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+    if (NOT VERSION_STRING)
+        # extract it from the existing generated header file
+        file(STRINGS "${CMAKE_CURRENT_SOURCE_DIR}/src/aws-cpp-sdk-core/include/aws/core/VersionConfig.h" __SDK_VERSION_LINE LIMIT_COUNT 1 REGEX "AWS_SDK_VERSION_STRING.*[0-9]+\\.[0-9]+\\.[0-9]+")
+        string(REGEX MATCH "([0-9]+\\.[0-9]+\\.[0-9]+)" VERSION_STRING "${__SDK_VERSION_LINE}")
+    endif ()
+
+    set(${resultVar} ${VERSION_STRING} PARENT_SCOPE)
+    message(STATUS "SDK project version: ${VERSION_STRING}")
```
Comment (sdavtaker) [inline cmake/sdk_versioning.cmake:19]: This should be DEBUG message but they are not available at the minimal cmake version we using

**Review by sbiscigl — APPROVED** [review]

**Review by SergeyRyabinin — APPROVED** [review]

**Pull request merged**